File: tests/integration/conditional-types-test/conditional-types-test.cpp
#include "conditional-types-test.h"

using namespace js;

template<typename T>
js::typed::StringOrNumber processValue(T value) {
    if (js::typeof_op(value) === "string"_S) {
            {
                        return value.toUpperCase();
            }
    }
    return js::number(42);
}

bool isString(js::any value) {
    return js::typeof_op(value) === "string"_S;
}

js::typed::Nullable<js::string> extractString(js::any value) {
    if (js::typeof_op(value) === "string"_S) {
            {
                        return value;
            }
    }
    return js::null;
}

template<typename T>
js::typed::Nullable<T> removeNull(js::any value) {
    if ((value !== js::null && value !== js::undefined)) {
            {
                        return value;
            }
    }
    return js::undefined;
}

auto main() {
    const js::any stringResult = processValue("hello"_S);
    const js::any numberResult = processValue(js::number(123));
    js::console.log("String result:"_S, stringResult);
    js::console.log("Number result:"_S, numberResult);
    const js::typed::StringOrNumber mixed = "test"_S;
    if (isString(mixed)) {
            {
                        js::console.log("Is string:"_S, mixed);
            }
    }
    const js::any extracted = extractString("hello"_S);
    if (extracted !== js::null) {
            {
                        js::console.log("Extracted string:"_S, extracted);
            }
    }
    const js::typed::Nullable<js::string> value = "not null"_S;
    const js::any nonNull = removeNull(value);
    if (nonNull !== js::undefined) {
            {
                        js::console.log("Non-null value:"_S, nonNull);
            }
    }
    const js::typed::Nullable<js::string> nullValue = js::null;
    const js::any result = removeNull(nullValue);
    js::console.log("Result from null:"_S, result);
}

File: tests/integration/conditional-types-test.ts
// Test for TypeScript conditional types
// Note: Since C++ doesn't have native conditional types, we transpile to
// the resolved types at compile time

// Function with conditional return type
// In TypeScript, the return type changes based on the input type
// In C++, we generate a template or use js::any
function processValue<T>(value: T): string | number {
  if (typeof value === "string") {
    return (value as any).toUpperCase();
  }
  return 42;
}

// Type guard function
function isString(value: any): boolean {
  return typeof value === "string";
}

// Utility function similar to TypeScript's Extract
function extractString(value: string | number | boolean): string | null {
  if (typeof value === "string") {
    return value;
  }
  return null;
}

// Utility function similar to NonNullable
function removeNull<T>(value: T | null | undefined): T | undefined {
  if (value !== null && value !== undefined) {
    return value;
  }
  return undefined;
}

// Test usage
function main() {
  // Test processValue with different types
  const stringResult = processValue("hello");
  const numberResult = processValue(123);

  console.log("String result:", stringResult);
  console.log("Number result:", numberResult);

  // Test type guard
  const mixed: string | number = "test";
  if (isString(mixed)) {
    console.log("Is string:", mixed);
  }

  // Test extract utility
  const extracted = extractString("hello");
  if (extracted !== null) {
    console.log("Extracted string:", extracted);
  }

  // Test removeNull utility
  const value: string | null = "not null";
  const nonNull = removeNull(value);
  if (nonNull !== undefined) {
    console.log("Non-null value:", nonNull);
  }

  // Test with actual null
  const nullValue: string | null = null;
  const result = removeNull(nullValue);
  console.log("Result from null:", result);
}

main();
